### PR TITLE
GPU: Implemented more depth buffer formats.

### PR DESCRIPTION
**Pull request by Subv**
This fixes the horizontal lines in Picross E, Cubic Ninja, Cave Story 3D and possibly others.

### CODE REVIEW

**Comment by Subv** [comment]
This behavior seems to be wrong according to some hwtests, but the hwtests may be wrong, I wasn't able to get the second fill engine to work, and some things just don't make sense at all.
https://gist.github.com/Subv/0ccbf7e48f87554e346f

**Comment by Subv** [comment]
Found the real problem, it was due to these games using the DEPTH24_STENCIL8 depth buffer format, which was not implemented

File: src/video_core/pica.h [hunk 1]
```diff
@@ -418,6 +418,13 @@ struct Regs {
             RGBA4    = 4,
         };
 
+        enum DepthFormat : u32 {
+            DEPTH16 = 0,
```
Comment (purpasmart96) [inline src/video_core/pica.h]: Names should be like "Depth16"  in enum"s with no underslashes

File: src/video_core/pica.h [hunk 2]
```diff
@@ -418,6 +418,13 @@ struct Regs {
             RGBA4    = 4,
         };
 
+        enum DepthFormat : u32 {
+            D16 = 0,
+        
+            D24 = 2,
```
Comment (neobrain) [inline src/video_core/pica.h]: Please align these at the `=`

File: src/video_core/rasterizer.cpp [hunk 3]
```diff
@@ -100,23 +100,66 @@ static u32 GetDepth(int x, int y) {
...
+        case registers.framebuffer.D16:
+        {
+            u32 stride = registers.framebuffer.width * 2;
+            return *(u16_le*)(depth_buffer + VideoCore::GetMortonOffset(x, y, 2) + coarse_y * stride);
+        }
+        case registers.framebuffer.D24:
+        {
+            u32 stride = registers.framebuffer.width * 3;
+            u8* address = depth_buffer + VideoCore::GetMortonOffset(x, y, 3) + coarse_y * stride;
+            return (address[2] << 16) | (address[1] << 8) | address[0];
```
Comment (neobrain) [inline src/video_core/rasterizer.cpp]: Functions in color.h akin to how color formats are en-/decoded should be introduced for this.

**Comment by neobrain** [comment]
If you feel like it, it would also be nice to have depth preview implemented in the framebuffer widget. For this purpose, what needs to be done is:
* `framebuffer_source_list` should have a new option called `Active depth buffer`
* `framebuffer_format_control` needs to have the depth formats added
* `GraphicsFramebufferWidget::Format` needs to have the depth formats added
* `GraphicsFramebufferWidget::OnUpdate` needs to implement the decoding logic appropriately (using the helper functions which should be added in color.h)

Since UI code only supports 8 bits per color component when displaying images, it would either be possible to show the depth buffer as 8-bit greyscale (i.e. sacrifizing the lower 16 bits of precision) OR to just visualize it by copying the 24 bits of depth into the RGB components directly.

**Comment by Subv** [comment]
Updated to include the viewing code for the framebuffer widget.

File: src/citra_qt/debugger/graphics_framebuffer.cpp [hunk 4]
```diff
@@ -184,6 +187,18 @@ void GraphicsFramebufferWidget::OnUpdate()
         break;
     }
 
+    case Source::DepthBuffer:
+    {
+        auto framebuffer = Pica::registers.framebuffer;
```
Comment (neobrain) [inline src/citra_qt/debugger/graphics_framebuffer.cpp]: should be a const reference (not sure why I went with a non-reference above)

File: src/citra_qt/debugger/graphics_framebuffer.cpp [hunk 5]
```diff
@@ -240,3 +279,19 @@ void GraphicsFramebufferWidget::OnUpdate()
...
+        case Format::RGB8:
+        case Format::D24:
+            return 3;
+        case Format::RGB5A1:
+        case Format::RGB565:
+        case Format::RGBA4:
+        case Format::D16:
+            return 2;
+    }
+}
```
Comment (neobrain) [inline src/citra_qt/debugger/graphics_framebuffer.cpp]: missing newline at EOF

File: src/citra_qt/debugger/graphics_framebuffer.h [hunk 6]
```diff
@@ -21,7 +21,8 @@ class GraphicsFramebufferWidget : public BreakPointObserverDock {
 
     enum class Source {
         PicaTarget = 0,
-        Custom = 1,
+        DepthBuffer = 1,
+        Custom = 2,
```
Comment (neobrain) [inline src/citra_qt/debugger/graphics_framebuffer.h]: Please align at `=`

File: src/core/hw/gpu.cpp [hunk 7]
```diff
@@ -81,9 +81,9 @@ inline void Write(u32 addr, const T data) {
             if (config.fill_24bit) {
                 // fill with 24-bit values
                 for (u8* ptr = start; ptr < end; ptr += 3) {
-                    ptr[0] = config.value_24bit_b;
+                    ptr[0] = config.value_24bit_r;
```
Comment (neobrain) [inline src/core/hw/gpu.cpp]: This is unrelated to the PR description, please move it to a separate PR.

File: src/video_core/color.h [hunk 8]
```diff
@@ -101,6 +101,33 @@ inline const Math::Vec4<u8> DecodeRGBA4(const u8* bytes) {
...
+inline const u32 DecodeD16(const u8* bytes) {
+    return *reinterpret_cast<const u16_le*>(bytes);
+}
+
+/**
+ * Decode a depth value stored in D24 format
+ * @param bytes Pointer to encoded source value
+ * @return Depth value as an u32
+ */
+inline const u32 DecodeD24(const u8* bytes) {
```
Comment (neobrain) [inline src/video_core/color.h]: Primitive types need not be marged `const` in return values.

(the rationale in returning `const Math::Vec2<u32>` is that the caller might be thinking that the return value is a reference and treat it as such, e.g. doing stupid stuff like `DecodeD24S8(..).Normalize()`)

**Comment by neobrain** [comment]
Added a bunch of comments, other than that LGTM

**Comment by Subv** [comment]
Updated

File: src/video_core/color.h [hunk 9]
```diff
@@ -153,4 +180,34 @@ inline void EncodeRGBA4(const Math::Vec4<u8>& color, u8* bytes) {
         (Convert8To4(color.g()) << 8) | (Convert8To4(color.b()) << 4) | Convert8To4(color.a());
 }
 
+/**
+ * Encode a depth value as D16 format
+ * @param value Source depth value to encode
+ * @param bytes Pointer where to store the encoded value
+ */
+inline void EncodeD16(u32 value, u8* bytes) {
+    *reinterpret_cast<u16_le*>(bytes) = value & 0xFFFF;
```
Comment (bunnei) [inline src/video_core/color.h]: What is the format of value? If its a 32-bit depth value, shouldn't it be value >> 16?

File: src/video_core/color.h [hunk 10]
```diff
@@ -153,4 +180,34 @@ inline void EncodeRGBA4(const Math::Vec4<u8>& color, u8* bytes) {
...
+}
+
+/**
+ * Encode a depth value as D24 format
+ * @param value Source depth value to encode
+ * @param bytes Pointer where to store the encoded value
+ */
+inline void EncodeD24(u32 value, u8* bytes) {
```
Comment (bunnei) [inline src/video_core/color.h]: Again, assuming value here is probably already a 24-bit depth value? Please document if so

File: src/video_core/color.h (hunk 9 above)
Comment (Subv) [inline src/video_core/color.h]: The value can be anything, but only 16 bits are encoded, this is only called with values with less or equal than 16 bits

File: src/video_core/color.h (hunk 10 above)
Comment (Subv) [inline src/video_core/color.h]: Yes, the value doesn't change formats between Getdepth and Setdepth calls

File: src/video_core/color.h (hunk 9 above)
Comment (bunnei) [inline src/video_core/color.h]: Then why not make value a u16 and document in the docstring that it expects a unsigned 16-bit depth component?

File: src/video_core/color.h (hunk 10 above)
Comment (bunnei) [inline src/video_core/color.h]: Again, same comment as above. Please put it in the docstring. When I see a function declaration like `EncodeD24(u32 value, u8* bytes)`, I automatically assume value is a 32-bit component.

File: src/video_core/color.h [hunk 11]
```diff
@@ -153,4 +180,34 @@ inline void EncodeRGBA4(const Math::Vec4<u8>& color, u8* bytes) {
...
+    bytes[0] = value & 0xFF;
+    bytes[1] = (value >> 8) & 0xFF;
+    bytes[2] = (value >> 16) & 0xFF;
+}
+
+/**
+ * Encode depth and stencil values as D24S8 format
+ * @param depth Source depth values to encode
```
Comment (bunnei) [inline src/video_core/color.h]: `24-bit source depth component to encode`

File: src/video_core/rasterizer.cpp [hunk 12]
```diff
@@ -100,23 +100,65 @@ static u32 GetDepth(int x, int y) {
...
     
     const u32 coarse_y = y & ~7;
-    u32 stride = registers.framebuffer.width * 2;
 
-    // Assuming 16-bit depth buffer format until actual format handling is implemented
-    return *(u16*)(depth_buffer + VideoCore::GetMortonOffset(x, y, 2) + coarse_y * stride);
+    switch (registers.framebuffer.depth_format) {
+        case registers.framebuffer.D16:
+        {
+            u32 stride = registers.framebuffer.width * 2;
```
Comment (bunnei) [inline src/video_core/rasterizer.cpp]: Can we add a `bytes_per_pixel` variable like we did with setting/getting the color values?

File: src/video_core/color.h (hunk 9 above)
Comment (Subv) [inline src/video_core/color.h]: Because then we'd need a cast in Setdepth and Getdepth for it, would that be OK?

**Comment by Subv** [comment]
Updated

**Comment by bunnei** [comment]
Squash please

**Comment by Subv** [comment]
Squashed

**Comment by bunnei** [comment]
LGTM
